### PR TITLE
feat(worker): Added logic to the worker to update inventory memberships.

### PR DESCRIPTION
**Pull request by jaliyaudagedara**

### CODE REVIEW

File: src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            _logger = logger;
+            _inventoryClient = inventoryClient;
+            _inventoryMembershipBuilder = inventoryMembershipBuilder;
+        }
+
+        public async Task<IEnumerable<QueueWorkItem<Target>>> ProcessItem(QueueWorkItem<Target> workItem,
+            IServiceScope scope,
+            CancellationToken cancellationToken = new CancellationToken())
+        {
+            Target target = workItem?.Message?.Resource?.Data;
```
Comment (sachintana) [inline src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs]: Suggestion: `target` can be a meaningful name such as `inventoryTarget`.

File: src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            {
+                throw ExceptionFactory.CreateException($"{path} Should not be null or empty", HttpStatusCode.BadRequest);
+            }
+
+            return Guid.Parse(valueObject.ToString());
+        }
+
+        private static void ValidateIdentifiers(Guid inventoryId, Guid organizationId, object setId)
+        {
+            if (inventoryId == default)
```
Comment (sachintana) [inline src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs]: Can you use `Validate.IsNotNullOrDefault` could be more handy.

File: src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+            if (target == null)
+            {
+                _logger.LogWarning($"'{nameof(target)}' is null.", workItem);
+                return workItem.ConvertToEnumerable();
+            }
+
+            _logger.LogInformation($"Processing target with UniqueId: '{target.Metadata.UniqueId}'.", target);
+
+            Guid organizationId = GetValue(target.Id, "OrganizationId");
```
Comment (sachintana) [inline src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs]: Can use `IdentifierExtensions` and `ObjectExtensions` in core framework.

File: src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+            if (inventory == null)
+            {
+                throw ExceptionFactory.CreateException($"Could not find inventory {inventoryId}: {organizationId}", HttpStatusCode.BadRequest);
+            }
+
+            InventoryMembership membership = await _inventoryClient
+                                                     .GetInventoryMembershipByUniqueId(target.Metadata.UniqueId, organizationId, cancellationToken);
+
+            if (operationType.Equals("Created", StringComparison.OrdinalIgnoreCase)
```
Comment (sachintana) [inline src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs]: Can `operationType ` ever be null?

File: src/Cosmos.Inventories.MembershipManager.WorkApi/Builders/InventoryMembershipBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    public class InventoryMembershipBuilder : IInventoryMembershipBuilder
+    {
+        private readonly InventoryMembership _inventoryMembership;
+
+        public InventoryMembershipBuilder()
+        {
+            _inventoryMembership = new InventoryMembership();
+        }
+
+        public IInventoryMembershipBuilder WithIdentifier(InventoryMembershipIdentifier identifier)
```
[the remark this answers is not in the document]
Comment (jaliyaudagedara) [reply src/Cosmos.Inventories.MembershipManager.WorkApi/Builders/InventoryMembershipBuilder.cs]: I'd love to, but more focus needs to be put on the implementation of endpoint for getting the inventories with set.
I have created a ticket: https://thlonline.atlassian.net/browse/COSFLEETOP-2605 to address this later.

File: src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        public InventoryMembershipManagerTask(ILogger<InventoryMembershipManagerTask> logger, 
+            IInventoryClient inventoryClient,
+            IInventoryMembershipBuilder inventoryMembershipBuilder)
+        {
+            _logger = logger;
+            _inventoryClient = inventoryClient;
+            _inventoryMembershipBuilder = inventoryMembershipBuilder;
+        }
+
+        public async Task<IEnumerable<QueueWorkItem<Target>>> ProcessItem(QueueWorkItem<Target> workItem,
```
[the remark this answers is not in the document]
Comment (jaliyaudagedara) [reply src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs]: See above

File: src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs (hunk 1 above)
Comment (jaliyaudagedara) [reply src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs]: This should be fine

File: src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs (hunk 2 above)
Comment (jaliyaudagedara) [reply src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs]: Yes, of course!

File: src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs (hunk 3 above)
Comment (jaliyaudagedara) [reply src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs]: Will check

File: src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs (hunk 4 above)
Comment (jaliyaudagedara) [reply src/Cosmos.Inventories.MembershipManager.WorkApi/InventoryMembershipManagerTask.cs]: Shouldn't be.

**Pull request merged**